CefSharp.Wpf/Internals/MouseAdjustor.cs

```diff
@@ -0,0 +1,132 @@
+
+        public System.Windows.Point Update(Rect originalRect, Rect viewRect)
+        {
+            int x = originalRect.X,
+                prevX = originalRect.X,
+                y = originalRect.Y,
+                prevY = originalRect.Y,
+                xOffset = 0,
+                yOffset = 0;
+
+            // If popup goes outside the view, try to reposition origin
+            if (originalRect.X + originalRect.Width > viewRect.Width)
+            {
+                x = viewRect.Width - originalRect.Width;
+                xOffset = prevX - x;
+            }
+
+            if (originalRect.Y + originalRect.Height > viewRect.Height)
+            {
+                y = y - originalRect.Height - 20;
+                yOffset = prevY - y;
+            }
+
+            // If x or y became negative, move them to 0 again
+            if (x < 0)
+            {
+                x = 0;
+                xOffset = prevX;
+            }
+
+            if (y < 0)
+            {
+                y = 0;
+                yOffset = prevY;
+            }
+
+            if (x != prevX || y != prevY)
+            {
+                this.xOffset = xOffset;
+                this.yOffset = yOffset;
+
+                Rect teleportingRect = new Rect(x, y, x + originalRect.Width, y + originalRect.Height);
+
+                this.originalRect = originalRect;
+                this.teleportingRect = teleportingRect;
+
+                if (this.originalRect.Y < this.teleportingRect.Y + this.teleportingRect.Height)
+                {
+                    var newY = this.teleportingRect.Y + this.teleportingRect.Height;
+                    this.originalRect = new Rect(originalRect.X, newY, originalRect.Width, originalRect.Y + originalRect.Height - newY);
+                }
+            }
+
+            return new System.Windows.Point(x, y);
+        }
+
+        /// <summary>
+        /// Resets the offsets and original-rect.
+        /// </summary>
+        public void Reset()
+        {
+            this.xOffset = 0;
+            this.yOffset = 0;
+
+            this.originalRect = new Rect();
+            this.originalRect = new Rect();
+        }
+
+        /// <summary>
+        /// Adjusts the mouse-coordinates when the popup is visible.
+        /// </summary>
+        /// <param name="point">The original point.</param>
+        /// <returns>The adjusted point if needed, else the original point.</returns>
+        public Point GetAdjustedMouseCoords(System.Windows.Point point)
+        {
+            return !this.IsInsideOriginalRect(point) && IsInsideTeleportingRect(point)
+                ? new Point((int)point.X + this.xOffset, (int)point.Y + this.yOffset)
+                : new Point((int)point.X, (int)point.Y);
+        }
+
+        /// <summary>
+        /// Checks if the given point is inside the original-rect.
+        /// </summary>
+        /// <param name="point">The point.</param>
+        /// <returns>A boolean.</returns>
```

Comment:
Ideally `returns` should tell you about the meaning of the boolean.

e.g. Returns true if the is inside original rectangle otherwise returns false.

Reply:
Done.